File: ChatGPT_BP_Cpp/CWE-476/476_1b_BP.cpp
#include <iostream>
#include <cstdlib>  // for std::malloc, std::free
#include <exception>

struct person {
    int status;
    // Add other fields here, such as name, age, etc.
};

// Custom exception class for memory allocation failure
class MemoryAllocationException : public std::exception {
public:
    const char* what() const noexcept override {
        return "Memory allocation failed.";
    }
};

// Secure memory allocation method
person* allocate_person() {
    // Allocate memory for a person struct
    person* new_person = static_cast<person*>(std::malloc(sizeof(person)));
    
    // Null pointer check for secure coding practices
    if (new_person == nullptr) {
        throw MemoryAllocationException(); // Throw custom exception if memory allocation fails
    }
    
    return new_person;
}

// Function to initialize the person struct
void initialize_person(person* p) {
    if (p == nullptr) {
        std::cerr << "Null pointer error: Cannot initialize person." << std::endl;
        return;
    }
    // Initialize fields (avoiding potential buffer overflows)
    p->status = 0;
    // Initialize other fields with safe operations if needed
}

// Function to deallocate the memory of person struct
void deallocate_person(person* p) {
    if (p != nullptr) {
        std::free(p);  // Safely free the allocated memory
        p = nullptr;   // Avoid dangling pointer
    }
}

int main() {
    person* new_person = nullptr;

    try {
        // Allocate and initialize person struct
        new_person = allocate_person();
        initialize_person(new_person);

        // Example usage of the person struct
        std::cout << "Person status: " << new_person->status << std::endl;

    } catch (const MemoryAllocationException& e) {
        // Handle memory allocation failure
        std::cerr << e.what() << std::endl;
        return EXIT_FAILURE;
    } catch (const std::exception& e) {
        // Catch any other exceptions
        std::cerr << "An error occurred: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }

    // Deallocate memory for person struct
    deallocate_person(new_person);

    return EXIT_SUCCESS;
}

File: GitHubCopilot_BP_Java/CWE-476/476_1b_BP.java
package GitHubCopilot_BP_Java.CWE-476;

import java.nio.ByteBuffer;

class Person {
    int status;

    public Person() {
        this.status = 0;
    }
}

public class Main {
    public static void main(String[] args) {
        ByteBuffer buffer = null;
        Person person = null;

        try {
            // Memory Allocation
            buffer = ByteBuffer.allocateDirect(1024); // Allocate 1024 bytes of memory

            // Initialize Person object
            person = new Person();

            // Null Pointer Check
            if (person == null) {
                throw new NullPointerException("Memory allocation for person failed.");
            }

            // Set status field
            person.status = 0;

            // Data Handling
            System.out.println("Person status: " + person.status);

        } catch (Exception e) {
            // Exception Handling
            System.err.println("An error occurred: " + e.getMessage());
            e.printStackTrace();
        } finally {
            // Memory Deallocation
            buffer = null; // Nullify the buffer reference for garbage collection
            person = null; // Nullify the person reference for garbage collection
        }
    }
}